nbformat/v4/nbformat.v4.schema.json

```diff
@@ -76,6 +76,40 @@
                       },
                       "additionalProperties": true
                     }
+                },
+                "environments": {
+                    "description": "Collection of environments in which this notebook was authored and executed",
+                    "type": "array",
+                    "item": {
+                        "description": "Specification of an environment in which this notebook was authored and executed",
+                        "type": "object",
+                        "anyOf": [
+                            {"required": ["type", "source"]},
+                            {"required": ["type", "spec"]}
+                        ],
+                        "properties": {
+                            "type": {
+                                "description": "The type of environment specification, preferably a mimetype including a version number (e.g., application/vnd.docker.dockerfile.v1)",
+                                "type": "string"
+                            },
+                            "source": {
+                                "description": "URI of the environment specification stored external to the notebook document",
```

Comment:
I'd like to add notes for which keys are mutually exclusive, e.g. source should state that it is mutually exclusive with spec and vice versa.